examples/5_secret_storage.py

```diff
@@ -0,0 +1,419 @@
+        except DeleteNonExistentError:
+            # refresh cached value
+            curr_info = await self.get(team, namespace, entry_key)
+            return None  # was already deleted. return None.
+        return res
+
+    async def get(
+        self, team: str, namespace: str, entry_key: str
+    ) -> keybase1.KVGetResult:
+        res = await self.kvstore.get(team, namespace, entry_key)
+        info = json.loads(res.entry_value) if res.entry_value != "" else None
+        self.update_cache(entry_key, info, res.revision)
+        return res
+
+    async def list_entrykeys(
+        self, team: str, namespace: str
+    ) -> keybase1.KVListEntryResult:
+        res = await self.kvstore.list_entrykeys(team, namespace)
+        return res
+
+
+def bytes_to_str(x):
+    return b64encode(x).decode("utf-8")
+
+
+def str_to_bytes(x):
+    return b64decode(x.encode("utf-8"))
+
+
+class SecretKeyKVStoreClient:
+    """
+        A SecretKeyKVStoreClient that hides the entryKeys from Keybase servers.
+        It does so by HMACing entryKeys using a per-(team, namespace) secret,
+        and storing the HMAC instead of the plaintext entryKey. This approach
+        does not handle any secret rotation, and does not expect the secret to
+        change.
+
+        The plaintext entryKey is stored in it's corresponding JSON entryValue
+        under the key "_key" to enable listing.
+
+        This approach does not hide memory access patterns. Also, Keybase
+        servers prevent a removed team member from continuing to access a team's
+        data, but if that were somehow bypassed*, a former team member who still
+        knows the HMAC secret can check for the presence of specific entryKeys
+        (*but you probably have bigger issues to deal with in that case...).
+    """
+
+    KEY_KEY = "_key"
+    SECRET_KEY = "_secret"
+    SECRET_NUM_BYTES = 32
+
+    def __init__(self, bot):
+        # secrets = {team: {namespace: secret}}
+        self.secrets: Dict[str, Dict[str, bytes]] = {}
+        self.kvstore: KVStoreClient = bot.kvstore
+
+    async def load_secret(self, team, namespace) -> bytes:
+        if team not in self.secrets or namespace not in self.secrets[team]:
+            secret = secrets.token_bytes(self.SECRET_NUM_BYTES)
+            try:
+                # we don't expect self.SECRET_KEY's revision > 0
+                await self.kvstore.put(
+                    team, namespace, self.SECRET_KEY, bytes_to_str(secret), revision=1
+                )
+            except RevisionError:
+                res: keybase1.KVGetResult = await self.kvstore.get(
+                    team, namespace, self.SECRET_KEY
+                )
+                secret = str_to_bytes(res.entry_value)
+            if team not in self.secrets:
+                self.secrets[team] = {}
+            self.secrets[team][namespace] = secret
+        return self.secrets[team][namespace]
+
+    async def hmac_key(self, team, namespace, entry_key) -> str:
```

Comment:
NAB: it's probably a better practice to call this method something like `hidden_key`. it's only incidental that it happens to do an hmac to achieve those ends.